rust/lance-encoding/src/encodings/physical/bitpacking.rs

```diff
@@ -349,29 +363,47 @@ fn unpack_out_of_line<T: ArrowNativeType + BitPacking>(
         .div_ceil(data.bits_per_value as usize);
     let compressed_words = data.data.borrow_to_typed_slice::<T>();
 
-    let num_chunks = data.num_values as usize / words_per_chunk;
-    debug_assert_eq!(data.num_values as usize % words_per_chunk, 0);
+    let num_whole_chunks = num_values / ELEMS_PER_CHUNK as usize;
+    let tail_values = num_values % ELEMS_PER_CHUNK as usize;
+    let expected_full_words = num_whole_chunks * words_per_chunk;
+    let expected_new_len = expected_full_words + tail_values;
+    let tail_is_raw = tail_values > 0 && compressed_words.len() == expected_new_len;
 
-    // This is slightly larger than num_values because the last chunk has some padding, we will truncate at the end
-    #[allow(clippy::uninit_vec)]
-    let mut decompressed = Vec::with_capacity(num_chunks * ELEMS_PER_CHUNK as usize);
     #[allow(clippy::uninit_vec)]
-    unsafe {
-        decompressed.set_len(num_chunks * ELEMS_PER_CHUNK as usize);
-    }
+    let mut decompressed = Vec::with_capacity(num_values);
+    let mut chunk_buf = vec![T::from_usize(0).unwrap(); ELEMS_PER_CHUNK as usize];
 
-    for chunk_idx in 0..num_chunks {
+    for chunk_idx in 0..num_whole_chunks {
         let input_start = chunk_idx * words_per_chunk;
         let input_end = input_start + words_per_chunk;
-        let output_start = chunk_idx * ELEMS_PER_CHUNK as usize;
-        let output_end = output_start + ELEMS_PER_CHUNK as usize;
         unsafe {
             BitPacking::unchecked_unpack(
                 compressed_bits_per_value,
                 &compressed_words[input_start..input_end],
-                &mut decompressed[output_start..output_end],
+                &mut chunk_buf,
             );
         }
+        decompressed.extend_from_slice(&chunk_buf);
+    }
```

Comment:
Why this change?  Doesn't it introduce an extra copy?  I have to decompress into `chunk_buf` and then copy into `decompressed` instead of decompressing directly into `decompressed`?

Reply:
I tried to refactor this part a bit to get rid of the unsafe. It's now reverted.